### PR TITLE
micro 1.0.3 (new formula)

### PR DESCRIPTION
**Pull request by zyedidia**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally prior to submission with `brew install <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [x] Does your submission pass `brew audit --new-formula <formula>` (after doing `brew install <formula>`)?

---

[Micro](https://github.com/zyedidia/micro) is a new terminal-based text editor.

I hope everything looks good. I was also wondering if I should doing anything extra to get bottles working or if the bot handles all that on its own.

I know the test is a little underwhelming, but I couldn't think of anything good to use to test it, and I looked at the nano formula which does the same thing for the test.

Thanks!

### CODE REVIEW

**Comment by apjanke** [comment]
Looks good to me.

Yeah, the bot handles everything to do with the bottles. 

And the minimal test is fine for an interactive program like this.

This is a good PR! Thanks for the submission. Normally Homebrew is not in favor of self-submissions, but this project looks popular/mature enough that I don't think it's a problem in this case.

Other maintainers: could I get someone else to chime in on whether this
```
date = `go run tools/build-date.go`.strip!
```
fits in with Homebrew conventions, and whether including a compile date has ramifications for reproducible builds that we care about?

Other than that, this looks good to go. 👍

**Comment by apjanke** [comment]
P.S. This is a nice looking program; gonna try it out myself.

File: Formula/micro.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  homepage "https://github.com/zyedidia/micro"
+  url "https://github.com/zyedidia/micro/archive/v1.0.3.tar.gz"
+  sha256 "dc9a693d9eedbbf5a0048a092172a56de6b2b608f7d1419274c12416a840876d"
+  head "https://github.com/zyedidia/micro.git"
+
+  depends_on "go" => :build
+
+  go_resource "github.com/gdamore/encoding" do
+    url "https://github.com/gdamore/encoding.git",
+    :revision => "b23993cbb6353f0e6aa98d0ee318a34728f628b9"
```
Comment (zmwangx) [inline Formula/micro.rb]: The current agreement among maintainers is that `:revision` should be aligned with the opening quote, i.e.,

```ruby
   go_resource "github.com/gdamore/encoding" do
     url "https://github.com/gdamore/encoding.git",
         :revision => "b23993cbb6353f0e6aa98d0ee318a34728f628b9"
```

**Comment by zmwangx** [comment]
@apjanke I have no problem with the date really, many programs record that by default, you just don't see it in the formula since it's hidden in `Makefile`, `configure.ac`, `Makefile.am`, `CMakeLists.txt` or whatever.

As for `.strip!`, I would say `.strip` or `.chomp` are better, just based on what we have already.

**Comment by zmwangx** [comment]
Updated earlier comment with code because somehow I submitted without completing the sentences... Sorry!

**Comment by DomT4** [comment]
Bonus points if anyone wants to politely ask upstream to consider adopting the Go vendoring system at some future point.

**Comment by zmwangx** [comment]
> Bonus points if anyone wants to politely ask upstream to consider adopting the Go vendoring system at some future point.

OP is the upstream 😆

**Comment by DomT4** [comment]
Oh, well, consider that a polite request to consider Go vendoring upstream then 😅.

File: Formula/micro.rb [hunk 2]
```diff
@@ -3,92 +3,95 @@
...
+    hash = `git rev-parse --short HEAD`.chomp
+    date = `go run tools/build-date.go`.chomp
+    system "go", "build", "-o", bin/"micro", "-ldflags", <<-EOS.undent, "./cmd/micro"
+      -X main.Version=#{version}
+      -X main.CommitHash=#{hash}
+      -X 'main.CompileDate=#{date}'
+      EOS
   end
 
   test do
```
Comment (zmwangx) [inline Formula/micro.rb]:
```ruby
assert_match version.to_s, shell_output("#{bin}/micro -version")
```

would be better.

**Comment by zyedidia** [comment]
I have fixed the problems mentioned by @zmwangx, although I have left the test block as is because micro actually prints the version with:

```
Version: 1.0.3
Commit hash: ...
Compile on ...
```

So the assert there doesn't really work.

**Comment by zmwangx** [comment]
Well, `assert_match` is not `assert_equal`. Since the output contains the version number, the former will succeed.

**Comment by zyedidia** [comment]
Oh I see, thanks for letting me know -- it should be fixed now. I don't use Ruby very often :).

Also about govendor, I am considering switching to it, but I'm probably going to need some time to get that all sorted out.

**Comment by zmwangx** [comment]
> Oh I see, thanks for letting me know -- it should be fixed now.

No problem.

> Also about govendor, I am considering switching to it, but I'm probably going to need some time to get that all sorted out.

Thanks, no rush.

**Comment by zyedidia** [comment]
Is getting micro to use govendor necessary before this formula can be merged?

**Comment by zmwangx** [comment]
> Is getting micro to use govendor necessary before this formula can be merged?

No! I'm happy with the current state, but want to make sure @apjanke is happy too.

**Comment by DomT4** [comment]
> Is getting micro to use govendor necessary before this formula can be merged?

Nope. Just a nice to have in future as `go_resource`s are an imperfect solution to dependency pinning, and now Go themselves provide a better solution through vendoring & tools like `godep`, `govendor`, `gpm`, etc that's a better system for everyone.

**Comment by apjanke** [comment]
Looks like concerns have been addressed. Are we good to merge here?

**Comment by zmwangx** [comment]
👍

File: Formula/micro.rb [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    mkdir_p buildpath/"src/github.com/zyedidia"
+    ln_s buildpath, buildpath/"src/github.com/zyedidia/micro"
+    Language::Go.stage_deps resources, buildpath/"src"
+    hash = `git rev-parse --short HEAD`.chomp
+    date = `go run tools/build-date.go`.chomp
+    system "go", "build", "-o", bin/"micro", "-ldflags", <<-EOS.undent, "./cmd/micro"
+      -X main.Version=#{version}
+      -X main.CommitHash=#{hash}
+      -X 'main.CompileDate=#{date}'
+      EOS
```
Comment (zmwangx) [inline Formula/micro.rb]: Oops, `EOS` should be aligned with `system`. I'll address that when merging.

**Comment by zmwangx** [comment]
Merged as 4de6bf7. Thank you for your contribution to Homebrew!

**Pull request closed without merging**